### ISSUE TITLE
Builds with ament_python broken with sphinx >= 7.0.0

### ISSUE THREAD

**Issue by oysstu** [colcon/colcon-core#556]
The following way of obtaining a list of available commands from setup.py is broken on sphinx v7.0.0 and later.
https://github.com/colcon/colcon-core/blob/99c03262ad7edd4af5954c424c9248e0c4fe7307/colcon_core/task/python/build.py#L170-L185

~~This occurs because the setuptools integration is deprecated:~~
~~https://github.com/sphinx-doc/sphinx/issues/9595#issuecomment-1135018951~~

The end result is that ament_python continues as normal, but installs the package as an egg without the egg-info. 
https://github.com/colcon/colcon-core/blob/99c03262ad7edd4af5954c424c9248e0c4fe7307/colcon_core/task/python/build.py#L92

When building ros, the build will fail on ament_cmake_core because it cannot find ament_package. Egg modules are not imported even if the site-packages directory is in the python-path. It's possible to add something like this to make the egg import

```python
import site
import os
site.addsitedir(os.path.split(__file__)[0])
```

However, there's further breakage later in the build process in cmake, as it expects template files in a certain location (and not inside the compressed egg).

**Comment by 14mRh4X0r** [comment]
Funnily enough, this actually seems to be a bug with Sphinx. I've filed a PR with them: sphinx-doc/sphinx#11418

**Comment by oysstu** [comment]
Resolved in sphinx 7.0.1
